kolibri/content/api.py

```diff
@@ -38,17 +45,109 @@ def title_description_filter(self, queryset, value):
+                                                  WHERE NOT (incomplete_log.progress < 1 AND incomplete_log.content_id = incomplete_node.content_id)
+                                                  AND complete_log.user_id = %s
+                                                  AND incomplete_log.user_id = %s
+                                                  AND complete_log.progress = 1
+                                                  AND complete_node.rght = incomplete_node.lft - 1
+                                                  AND complete_log.content_id = complete_node.content_id''', [value, value])
+        else:
+            summary_logs = ContentSummaryLog.objects.filter(user=value).exclude(progress=1)
+
+        content_ids = summary_logs.values_list('content_id', flat=True)
+        unfinished_nodes = queryset.filter(content_id__in=list(content_ids[:10]))
+
+        return unfinished_nodes
+
+    def filter_popular(self, queryset, value):
+        """
+        Recommend content that is popular with all users.
+
+        :param queryset: all content nodes for this channel
+        :param value: id of currently logged in user, or none if user is anonymous
+        :return: 10 most popular content nodes
+        """
+
+        if ContentSessionLog.objects.count() < 50:
+            # return 25 random content nodes if not enough session logs
+            pks = queryset.values_list('pk', flat=True).exclude(kind__in=['topic', ''])
+            count = min(pks.count(), 25)
+            return queryset.filter(pk__in=sample(list(pks), count))
+
+        if self.data['channel']:  # filter by channel if available
+            session_logs = ContentSessionLog.objects.filter(channel_id=self.data['channel'])
+            cache_key = 'popular_for_{}'.format(self.data['channel_id'])
+            if cache.get(cache_key):
+                return cache.get(cache_key)
+        else:
+            session_logs = ContentSessionLog.objects.all()
+            cache_key = 'popular'
+            if cache.get(cache_key):
+                return cache.get(cache_key)
+
+        # get the most accessed content nodes
+        content_counts_sorted = session_logs.values_list('content_id', flat=True).annotate(Count('content_id')).order_by('-content_id__count')
+        most_popular = queryset.filter(content_id__in=list(content_counts_sorted[:10]))
+
+        # cache the popular results queryset for 10 minutes, for efficiency
+        cache.set(cache_key, most_popular, 60 * 10)
+        return most_popular
+
+    def filter_resume(self, queryset, value):
+        """
+        Recommend content that the user has recently engaged with, but not finished.
+
+        :param queryset: all content nodes for this channel
+        :param value: id of currently logged in user, or none if user is anonymous
+        :return: 10 most recently viewed content nodes
+        """
+
+        # if user is anonymous, return no nodes
+        if value == '':
```

Comment:
@jamalex @rtibbles I had to do this check because it seems that null was getting sent in the request to the server from the client, and then the server converted this to an empty string. But I think this is wrong, it should be converted to a None type.